src/csharp/Grpc.Core/Internal/AsyncCallBase.cs

```diff
@@ -172,7 +188,8 @@ protected bool ReleaseResourcesIfPossible()
             if (!disposed && call != null)
             {
                 bool noMoreSendCompletions = streamingWriteTcs == null && (halfcloseRequested || cancelRequested || finished);
-                if (noMoreSendCompletions && readingDone && finished && !receiveResponseHeadersPending)
+                bool readingFinished = readingDone || IsInErrorStateClientOnly;
```

Comment:
I'm now using 
`bool readingFinished = readingDone || (IsFinishedWithNonOkStatusClientOnly && streamingReadTcs == null);`
the `streamingReadTcs == null` is the check for no pending read although I'm not sure it is necessary since if there was an error the a new read would not have been started.